### ISSUE TITLE
Allow `Match` to accept regex

### ISSUE THREAD

**Issue by chielkunkels** [ReactTraining/react-router#4164]
Just getting to grips with RR4 - quite like it so far. One thing I was struggling with a bit, due to how routes are declared now, is how to render a component if it's NOT a match.

For example, I have an application which is behind a login wall, so I don't want to render the application UI until people are in.

Naïvely I initially had:

```jsx
<Match pattern="/" component={Main} />
<Match pattern="/login" component={Login} />
```

This of course went wrong due to `/login` also matching `/`. I tried searching for a way to render a component if the pattern did NOT match, but couldn't find anything in the docs. I imagine it would have looked something like this:

```jsx
<Match not pattern="/login" component={Main} />
<Match pattern="/login" component={Login} />
```

`not` might as well be `negative` or what have you. Sadly, such an option did not seem to exist. Next up I tried to just pass in a regex with a negative lookahead to see if this worked:

```jsx
<Match pattern={/^\/(?!login)/} component={Main} />
<Match pattern="/login" component={Login} />
```

Success! At least, in terms of the actual routes working. Now I get a warning from React about the prop being of an incorrect type:

```
Warning: Failed prop type: Invalid prop `pattern` of type `regexp` supplied to `Match`, expected `string`.
    in Match (created by App)
    in App
    in MatchProvider (created by StaticRouter)
    in Broadcast (created by LocationBroadcast)
    in LocationBroadcast (created by StaticRouter)
    in StaticRouter (created by History)
    in History (created by BrowserRouter)
    in BrowserRouter
    in Provider
```

I might be completely abusing the way this is supposed to work, but I couldn't find any other way to achieve this without actually redoing  my URL structure, which I wasn't too keen on.

**TL;DR**: Any chance we can also support regex as `pattern`? Happy to contribute with a pull request.

**Comment by akupila** [comment]
Does this work for you?
```js
<Match exactly pattern="/" component={Main} />
<Match pattern="/login" component={Login} />
```

**Comment by chielkunkels** [comment]
@akupila no, because I want the `Main` component to render on every route except `Login`. That's because it basically contains the general application UI.

**Comment by akupila** [comment]
Ah, of course. I'd imagine react-router can't (or shouldn't) help you here, simply render the main app or the login page depending on login state? Alternatively redirect to `/login` if not logged in and deeplinked to the app or vice versa?

**Comment by pshrmn** [comment]
It does: https://github.com/pillarjs/path-to-regexp#custom-match-parameters

With the current release, this would be solved using a `<Miss>`, but there are a number of issues with that component. The v4 API is still being worked on and there will be better ways to do this than to use a regex. The two most likely solutions are either using `<MatchGroup>` to define a set of `<Match>`es where only one will be rendered or `<Match>` will be modified to accept an array of patterns and you could use the wildcard pattern to match all non-login paths

**Comment by chielkunkels** [comment]
@pshrmn hmm, I don't really see how that helps exactly? But you're right, it's probably better to render components based on logged in state then, thanks!

**Comment by taurose** [comment]
Seems related: #3906

**Comment by AlekzZz** [comment]
what a great example of "type correctness !== program correctness"
